File: frontend/src/types/coms.ts
/**
 * Coms (Communications) Type Definitions
 * Production communications system - channels, messages, voice, presence
 */

// ============================================================================
// ENUMS
// ============================================================================

export type ChannelType = 'dm' | 'group_chat' | 'voice' | 'text_and_voice';
export type ChannelScope = 'project' | 'global';
export type MessageType = 'text' | 'system' | 'file' | 'voice_note';
export type PresenceStatus = 'online' | 'away' | 'busy' | 'offline';
export type ChannelMemberRole = 'admin' | 'moderator' | 'member';

// ============================================================================
// CHANNEL TYPES
// ============================================================================

export interface ComsChannel {
  id: string;
  name: string;
  description: string | null;
  channel_type: ChannelType;
  scope: ChannelScope;
  project_id: string | null;
  icon: string | null;
  color: string | null;
  template_key: string | null;
  is_system_channel: boolean;
  visible_to_roles: string[];
  can_transmit_roles: string[];
  is_private: boolean;
  created_by: string | null;
  sort_order: number;
  created_at: string;
  updated_at: string;
  archived_at: string | null;
  // Computed
  unread_count: number;
  member_count: number;
  last_message?: ComsMessage | null;
}

export interface ChannelMemberInfo {
  id: string;
  user_id: string;
  username: string | null;
  full_name: string | null;
  avatar_url: string | null;
  role: ChannelMemberRole;
  can_transmit: boolean;
  is_muted: boolean;
  joined_at: string;
}

export interface ComsChannelWithMembers extends ComsChannel {
  members: ChannelMemberInfo[];
}

export interface ChannelCreateInput {
  name: string;
  description?: string;
  channel_type?: ChannelType;
  scope?: ChannelScope;
  project_id?: string;
  icon?: string;
  color?: string;
  visible_to_roles?: string[];
  can_transmit_roles?: string[];
  is_private?: boolean;
  template_key?: string;
}

export interface ChannelUpdateInput {
  name?: string;
  description?: string;
  icon?: string;
  color?: string;
  visible_to_roles?: string[];
  can_transmit_roles?: string[];
  is_private?: boolean;
}

// ============================================================================
// MESSAGE TYPES
// ============================================================================

export interface SenderInfo {
  id: string;
  username: string | null;
  full_name: string | null;
  avatar_url: string | null;
  production_role: string | null;
}

export interface ComsMessage {
  id: string;
  channel_id: string;
  sender_id: string;
  content: string;
  message_type: MessageType;
  attachments: any[];
  reply_to_id: string | null;
  edited_at: string | null;
  is_deleted: boolean;
  created_at: string;
  sender?: SenderInfo;
  reply_to?: ComsMessage;
}

export interface MessageCreateInput {
  content: string;
  message_type?: MessageType;
  attachments?: any[];
  reply_to_id?: string;
}

export interface MessagePage {
  messages: ComsMessage[];
  has_more: boolean;
  next_cursor: string | null;
}

// ============================================================================
// CHANNEL MEMBER TYPES
// ============================================================================

export interface ChannelMemberAddInput {
  user_id: string;
  role?: ChannelMemberRole;
  can_transmit?: boolean;
}

export interface ChannelMember {
  id: string;
  channel_id: string;
  user_id: string;
  role: ChannelMemberRole;
  can_transmit: boolean;
  is_muted: boolean;
  notifications_enabled: boolean;
  joined_at: string;
  username?: string | null;
  full_name?: string | null;
  avatar_url?: string | null;
}

// ============================================================================
// VOICE TYPES
// ============================================================================

export interface VoiceParticipant {
  id: string;
  user_id: string;
  username: string | null;
  full_name: string | null;
  avatar_url: string | null;
  production_role: string | null;
  is_transmitting: boolean;
  is_muted: boolean;
  is_deafened: boolean;
  peer_id: string | null;
  joined_at: string;
}

export interface VoiceJoinResponse {
  room_id: string;
  channel_id: string;
  ice_servers: RTCIceServer[];
  participants: VoiceParticipant[];
}

export interface VoiceRoom {
  id: string;
  channel_id: string;
  is_active: boolean;
  max_participants: number;
  started_at: string;
  ended_at: string | null;
  participants: VoiceParticipant[];
}

// ============================================================================
// PRESENCE TYPES
// ============================================================================

export interface UserPresence {
  user_id: string;
  username: string | null;
  full_name: string | null;
  avatar_url: string | null;
  status: PresenceStatus;
  status_message: string | null;
  current_channel_id: string | null;
  current_project_id: string | null;
  last_seen_at: string;
}

export interface ProjectPresence {
  project_id: string;
  users: UserPresence[];
  online_count: number;
  away_count: number;
}

export interface PresenceUpdateInput {
  status: PresenceStatus;
  status_message?: string;
  current_channel_id?: string;
  current_project_id?: string;
}

// ============================================================================
// READ RECEIPT TYPES
// ============================================================================

export interface UnreadCount {
  channel_id: string;
  channel_name: string;
  unread_count: number;
}

export interface UnreadCountsResponse {
  total_unread: number;
  channels: UnreadCount[];
}

// ============================================================================
// TEMPLATE TYPES
// ============================================================================

export interface ChannelTemplate {
  id: string;
  template_key: string;
  name: string;
  description: string | null;
  channel_type: ChannelType;
  icon: string | null;
  color: string | null;
  default_visible_to_roles: string[];
  default_can_transmit_roles: string[];
  sort_order: number;
  is_active: boolean;
}

export interface ApplyTemplatesRequest {
  template_keys: string[];
}

export interface ApplyTemplatesResponse {
  created_channels: ComsChannel[];
  skipped_templates: string[];
}

// ============================================================================
// LIST RESPONSE TYPES
// ============================================================================

export interface ChannelListResponse {
  channels: ComsChannel[];
  total: number;
}

// ============================================================================
// CONSTANTS
// ============================================================================

export const CHANNEL_TYPE_CONFIG: Record<ChannelType, { label: string; icon: string; color: string }> = {
  dm: { label: 'Direct Message', icon: 'message-square', color: '#6B7280' },
  group_chat: { label: 'Group Chat', icon: 'message-circle', color: '#3B82F6' },
  voice: { label: 'Voice Only', icon: 'mic', color: '#10B981' },
  text_and_voice: { label: 'Text & Voice', icon: 'radio', color: '#8B5CF6' },
};

export const PRESENCE_STATUS_CONFIG: Record<PresenceStatus, { label: string; color: string }> = {
  online: { label: 'Online', color: '#10B981' },
  away: { label: 'Away', color: '#F59E0B' },
  busy: { label: 'Busy', color: '#EF4444' },
  offline: { label: 'Offline', color: '#6B7280' },
};

export const CHANNEL_ICON_MAP: Record<string, string> = {
  megaphone: 'Megaphone',
  timer: 'Timer',
  camera: 'Camera',
  aperture: 'Aperture',
  building: 'Building',
  lock: 'Lock',
  'message-circle': 'MessageCircle',
  zap: 'Zap',
  'volume-2': 'Volume2',
  palette: 'Palette',
};

File: backend/app/schemas/coms.py
"""
Coms (Communications) Pydantic Schemas
Production communications system - channels, messages, voice, presence
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from datetime import datetime
from enum import Enum


# ============================================================================
# ENUMS
# ============================================================================

class ChannelType(str, Enum):
    DM = "dm"
    GROUP_CHAT = "group_chat"
    VOICE = "voice"
    TEXT_AND_VOICE = "text_and_voice"


class ChannelScope(str, Enum):
    PROJECT = "project"
    GLOBAL = "global"


class MessageType(str, Enum):
    TEXT = "text"
    SYSTEM = "system"
    FILE = "file"
    VOICE_NOTE = "voice_note"


class PresenceStatus(str, Enum):
    ONLINE = "online"
    AWAY = "away"
    BUSY = "busy"
    OFFLINE = "offline"


class ChannelMemberRole(str, Enum):
    ADMIN = "admin"
    MODERATOR = "moderator"
    MEMBER = "member"


# ============================================================================
# CHANNEL SCHEMAS
# ============================================================================

class ChannelBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    channel_type: ChannelType = ChannelType.TEXT_AND_VOICE
    icon: Optional[str] = None
    color: Optional[str] = None
    visible_to_roles: List[str] = []
    can_transmit_roles: List[str] = []
    is_private: bool = False


class ChannelCreate(ChannelBase):
    project_id: Optional[str] = None
    scope: ChannelScope = ChannelScope.PROJECT
    template_key: Optional[str] = None


class ChannelUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    icon: Optional[str] = None
    color: Optional[str] = None
    visible_to_roles: Optional[List[str]] = None
    can_transmit_roles: Optional[List[str]] = None
    is_private: Optional[bool] = None


class ChannelMemberInfo(BaseModel):
    id: str
    user_id: str
    username: Optional[str] = None
    full_name: Optional[str] = None
    avatar_url: Optional[str] = None
    role: str = "member"
    can_transmit: bool = True
    is_muted: bool = False
    joined_at: datetime


class Channel(ChannelBase):
    id: str
    scope: ChannelScope
    project_id: Optional[str]
    template_key: Optional[str]
    is_system_channel: bool
    created_by: Optional[str]
    sort_order: int
    created_at: datetime
    updated_at: datetime
    archived_at: Optional[datetime] = None
    # Computed fields
    unread_count: int = 0
    member_count: int = 0
    last_message: Optional["Message"] = None

    class Config:
        from_attributes = True


class ChannelWithMembers(Channel):
    members: List[ChannelMemberInfo] = []


# ============================================================================
# MESSAGE SCHEMAS
# ============================================================================

class MessageBase(BaseModel):
    content: str = Field(..., min_length=1)
    message_type: MessageType = MessageType.TEXT
    attachments: List[Any] = []
    reply_to_id: Optional[str] = None


class MessageCreate(MessageBase):
    pass


class MessageUpdate(BaseModel):
    content: str = Field(..., min_length=1)


class SenderInfo(BaseModel):
    id: str
    username: Optional[str] = None
    full_name: Optional[str] = None
    avatar_url: Optional[str] = None
    production_role: Optional[str] = None


class Message(MessageBase):
    id: str
    channel_id: str
    sender_id: str
    edited_at: Optional[datetime] = None
    is_deleted: bool = False
    created_at: datetime
    # Joined data
    sender: Optional[SenderInfo] = None
    reply_to: Optional["Message"] = None

    class Config:
        from_attributes = True


class MessagePage(BaseModel):
    messages: List[Message]
    has_more: bool
    next_cursor: Optional[str] = None


# ============================================================================
# CHANNEL MEMBER SCHEMAS
# ============================================================================

class ChannelMemberAdd(BaseModel):
    user_id: str
    role: ChannelMemberRole = ChannelMemberRole.MEMBER
    can_transmit: bool = True


class ChannelMemberUpdate(BaseModel):
    role: Optional[ChannelMemberRole] = None
    can_transmit: Optional[bool] = None
    is_muted: Optional[bool] = None
    notifications_enabled: Optional[bool] = None


class ChannelMember(BaseModel):
    id: str
    channel_id: str
    user_id: str
    role: str
    can_transmit: bool
    is_muted: bool
    notifications_enabled: bool
    joined_at: datetime
    # Joined user data
    username: Optional[str] = None
    full_name: Optional[str] = None
    avatar_url: Optional[str] = None

    class Config:
        from_attributes = True


# ============================================================================
# VOICE SCHEMAS
# ============================================================================

class VoiceJoinResponse(BaseModel):
    room_id: str
    channel_id: str
    ice_servers: List[Any]
    participants: List["VoiceParticipant"]


class VoiceParticipant(BaseModel):
    id: str
    user_id: str
    username: Optional[str] = None
    full_name: Optional[str] = None
    avatar_url: Optional[str] = None
    production_role: Optional[str] = None
    is_transmitting: bool = False
    is_muted: bool = False
    is_deafened: bool = False
    peer_id: Optional[str] = None
    joined_at: datetime

    class Config:
        from_attributes = True


class VoiceRoom(BaseModel):
    id: str
    channel_id: str
    is_active: bool
    max_participants: int
    started_at: datetime
    ended_at: Optional[datetime] = None
    participants: List[VoiceParticipant] = []

    class Config:
        from_attributes = True


# ============================================================================
# PRESENCE SCHEMAS
# ============================================================================

class PresenceUpdate(BaseModel):
    status: PresenceStatus
    status_message: Optional[str] = None
    current_channel_id: Optional[str] = None
    current_project_id: Optional[str] = None


class UserPresence(BaseModel):
    user_id: str
    username: Optional[str] = None
    full_name: Optional[str] = None
    avatar_url: Optional[str] = None
    status: PresenceStatus
    status_message: Optional[str] = None
    current_channel_id: Optional[str] = None
    current_project_id: Optional[str] = None
    last_seen_at: datetime

    class Config:
        from_attributes = True


class ProjectPresence(BaseModel):
    project_id: str
    users: List[UserPresence]
    online_count: int
    away_count: int


# ============================================================================
# READ RECEIPT SCHEMAS
# ============================================================================

class MarkReadRequest(BaseModel):
    message_id: Optional[str] = None  # If None, marks all as read


class UnreadCount(BaseModel):
    channel_id: str
    channel_name: str
    unread_count: int


class UnreadCountsResponse(BaseModel):
    total_unread: int
    channels: List[UnreadCount]


# ============================================================================
# TEMPLATE SCHEMAS
# ============================================================================

class ChannelTemplate(BaseModel):
    id: str
    template_key: str
    name: str
    description: Optional[str]
    channel_type: ChannelType
    icon: Optional[str]
    color: Optional[str]
    default_visible_to_roles: List[str]
    default_can_transmit_roles: List[str]
    sort_order: int
    is_active: bool

    class Config:
        from_attributes = True


class ApplyTemplatesRequest(BaseModel):
    template_keys: List[str]


class ApplyTemplatesResponse(BaseModel):
    created_channels: List[Channel]
    skipped_templates: List[str]  # Already exist


# ============================================================================
# LIST RESPONSE SCHEMAS
# ============================================================================

class ChannelListResponse(BaseModel):
    channels: List[Channel]
    total: int


# Update forward references
Channel.model_rebuild()
Message.model_rebuild()
VoiceJoinResponse.model_rebuild()
